src/app/mesh/qgsmaptooleditmeshframe.cpp

```diff
@@ -2047,6 +2080,40 @@ void QgsMapToolEditMeshFrame::selectContainedByGeometry( const QgsGeometry &geom
   setSelectedVertices( selectedVertices.values(), behavior );
 }
 
+void QgsMapToolEditMeshFrame::applyZValueFromProjectTerrainOnSelectedVertices()
+{
+  if ( !mZValueWidget )
+    return;
+
+  if ( mSelectedVertices.isEmpty() )
+    return;
+
+  const QgsAbstractTerrainProvider *terrainProvider = QgsProject::instance()->elevationProperties()->terrainProvider();
+  const QgsCoordinateTransform transformation = QgsCoordinateTransform( mCurrentLayer->crs(), terrainProvider->crs(), QgsProject::instance() );
+
+  QList<double> zValues;
+  zValues.reserve( mSelectedVertices.count() );
+
+  for ( QMap<int, SelectedVertexData>::iterator it = mSelectedVertices.begin(); it != mSelectedVertices.end(); ++it )
+  {
+    const QgsPoint vertex = mapVertex( it.key() );
+    QgsPointXY point;
+
+    try
+    {
+      point = transformation.transform( vertex.x(), vertex.y() );
+    }
+    catch ( const QgsCsException & )
+    {
+      point = QgsPointXY( vertex.x(), vertex.y() );
```

Comment:
I doubt this is the correct thing to do. That will lead to getting garbage Z just afterwards. Maybe insert ``std::numeric_limits<double>::quiet_NaN()`` in zValues ... ? But that might cause downstream issues...